### PR TITLE
[ci] Try to use GCP Bazel cache for more jobs

### PR DESCRIPTION
**Pull request by jwnrt**
## Summary

This change lets more jobs _read_ from the cache, and also allows more jobs to _write_ to the cache when running for the `master` branch.

It's being tested in [this Azure run][azure-run] which will write to the cache. A second run will need to be triggered to see if we actually read from the cache.

## Details

This change:

1. Switches usage of `./bazelisk.sh` to `ci/bazelisk.sh` in pipelines.
   * This script connects Bazel to our GCP bucket containing a Bazel cache.
   * This allows CI to read from the cache.
   * If the `GCP_BAZEL_CACHE_KEY` variable is set, it also writes to the cache.
2. Adds downloads for the file containing cache write credentials to more jobs.
   * This file is stored encrypted in Azure and cannot be downloaded by PRs from forks.
   * The location to this file is automatically stored in the `GCP_BAZEL_CACHE_KEY_SECUREFILEPATH` environment variable.
   * The file is only downloaded for the `master` branch, however I have changed this for [this test run][azure-run] to show that the cache can be written to.
   * In summary: having this downloaded allows the master branch to write to the cache.

[azure-run]: https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=138706&view=results

### CODE REVIEW

**Comment by pamaury** [comment]
Thanks @jwnrt for looking at this, this looks like a great idea. It seems that now we will have many instances of
```yaml
  - task: DownloadSecureFile@1
    condition: eq(variables['Build.SourceBranchName'], 'master')
    name: GCP_BAZEL_CACHE_KEY
    inputs:
      secureFile: "bazel_cache_gcp_key.json"
```
which is quite error-prone and very specific to our workflow. What do you think about having a yaml file that we can include so we can do instead something like:
```yaml
- template: ci/use-bazel-remove-cache.yaml
```
and maybe add a parameter to specific whether access is just read, or read-write?

**Comment by jwnrt** [comment]
I've extracted this step into a template but haven't added an r/rw parameter since these credentials are only for writing. The GCP bucket is always read if using the `ci/bazelisk.sh`.

I've tried to make this clearer by giving the template the name `load-bazel-cache-write-creds.yml` so it's more obvious they just for writing.

**Comment by jwnrt** [comment]
Oh also I should mention that even with this change, I don't think the Bazel cache is working in CI at all.

Before this PR, the `sw_build` job was set up correctly to use the cache (it had the GCP credentials to write and it was using `ci/bazelisk.sh` to read) but as far as I can tell has always rebuilt.

I tried to set up a local environment to read from the cache too, but no success. These are the (convoluted) steps I used (thanks to @nbdd0121 for helping with this):

1. Added the following to `~/.bazelrc` to replicate what `ci/bazelisk.sh` builds:
```
build --remote_cache=https://storage.googleapis.com/opentitan-bazel-cache
build --remote_upload_local_results=false
build --remote_default_exec_properties=OSVersion="Ubuntu 20.04.6 LTS"
```
2. Set `HOME="/root"` - this is what the CI uses and the `HOME` env var is non-hermetic, so may have influenced the cache?
3. Set `PATH=/tools/verilator/v4.210/bin:/tools/verible/bin:/root/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin` which is what `sw_build` uses - we extracted this by running `env` in [this PR's CI run](https://github.com/lowRISC/opentitan/pull/20841).
4. Execute the same build command that `sw_build` does.

We did not observe the cache being used at all. There must be something else contributing to the cache that we didn't account for.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -214,14 +205,13 @@ jobs:
...
       #   environment does not have access to Vivado.
-      export GCP_BAZEL_CACHE_KEY=$(bazelCacheGcpKeyPath)
       TARGET_PATTERN_FILE=target_pattern.txt
       echo //... > "${TARGET_PATTERN_FILE}"
       echo -//quality/... >> "${TARGET_PATTERN_FILE}"
       echo -//sw/otbn/crypto/... >> "${TARGET_PATTERN_FILE}"
       echo -//third_party/riscv-compliance/... >> "${TARGET_PATTERN_FILE}"
       echo -//hw:all >> "${TARGET_PATTERN_FILE}"
-      ./bazelisk.sh cquery \
+      ci/bazelisk.sh cquery \
```
Comment (pamaury) [inline azure-pipelines.yml:214]: Since this is just a query, I am not sure that it makes a difference whether it uses a the cache or not. Also, if I recall correctly, `ci/bazellisk.sh` can print things that could make the output unusuable for queries (I might recall incorrectly).

Comment (jwnrt) [reply azure-pipelines.yml:214]: You're right about it printing output when it shouldn't - I've moved those to `stderr` so they shouldn't interfere.

Do you think there's much of a penalty of using `ci/bazelisk.sh` for everything in CI? It might be easier to be consistent and use it for everything. I don't think everybody knows you need to use that script to get caching, so using it consistently in CI may encourage others to use it when adding things?

Comment (jwnrt) [reply azure-pipelines.yml:214]: Oh, I do see that our custom `bazelisk.sh` commands (like outquery) aren't working when run through `ci/bazelisk.sh`, but I'm not sure why :thinking:

Maybe I will move those back to `./bazelisk.sh` if I can't work out what's wrong.

Comment (pamaury) [reply azure-pipelines.yml:214]: I tend to agree that using `ci/bazelisk.sh` consistently would be better. This script calls into `bazelisk.sh` so in principle our custom commands should work?

Comment (jwnrt) [reply azure-pipelines.yml:214]: Ah, it's because `./bazelisk.sh` only checks `$1` for the command (e.g. `outquery`) while `ci/bazelisk.sh` adds some flags before your arguments (i.e. `--bazelrc=/tmp/...`).

We could change the order that `ci/bazelisk.sh` passes its flags or give `./bazelisk.sh` better argument parsing.

Comment (jwnrt) [reply azure-pipelines.yml:214]: I've made `./bazelisk.sh` correctly handle flags that have to come before the subcommand (like `--bazelrc=...`) but it's using some fairly ugly Bash. It's valid bash AFAICT (shellcheck is okay with it) but let me know if you think it's acceptable.

File: ci/bazelisk.sh [hunk 2]
```diff
@@ -15,12 +15,13 @@ echo "Running bazelisk in $(pwd)."
 
 # An additional bazelrc must be synthesized to specify precisely how to use the
 # GCP bazel cache.
+GCP_CREDS_FILE="$GCP_BAZEL_CACHE_KEY_SECUREFILEPATH"
```
Comment (pamaury) [inline ci/bazelisk.sh:18]: I am assuming that the `_SECUREFILEPATH` is added by `DownloadSecureFile` to the name? It's not really clear in the documentation https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/download-secure-file-v1?view=azure-pipelines

Comment (jwnrt) [reply ci/bazelisk.sh:18]: `DownloadSecureFile` will store the path to the file in a variable called `$(GCP_BAZEL_CACHE_KEY.secureFilePath)`, and Azure also has a rule that you can access `$(AZURE_VARIABLES.WITH.SCOPES)` by replacing `.`s with `_`s

**Review by pamaury — APPROVED** [review]
Thanks @jwnrt, this looks good to me. I am no bash expert though so I might have missed something.

**Comment by jwnrt** [comment]
Thanks for reviewing, I asked @nbdd0121 to look over the Bash and we simplified it a bit more
